### ISSUE TITLE
this access should be allowed in property initializors

### ISSUE THREAD

**Issue by doug-wade** [eslint/eslint#8271]
Currently, the stage 2 property initializer syntax is being identified as violating `no-invalid-this`, and I don't think it should.  The new syntax is [stage 2](https://github.com/tc39/proposal-class-public-fields), and parses with `babel-eslint`, so I think it should be supported by eslint.  The [spec](https://tc39.github.io/proposal-class-public-fields/#[[construct]]) indicates that `this` should be bound to the [instance](https://tc39.github.io/ecma262/#sec-ordinarycreatefromconstructor), and accordingly it compiles in the [babel online repl](https://babeljs.io/repl/#?babili=false&evaluate=true&lineWrap=false&presets=es2015%2Creact%2Cstage-2&targets=&browsers=&builtIns=false&code=class%20Foo%20%7B%0A%20%20methodName%20%3D%20()%20%3D%3E%20%7B%0A%20%20%20%20%2F%2F%20nothing%0A%20%20%7D%0A%7D) to close over `this` from the instance, so I think it is valid to use the `this` keyword inside the property.  I've filed https://github.com/eslint/eslint/pull/8272 with a failing unit test to reproduce this issue.

**Tell us about your environment**

* **ESLint Version:**
3.17.1

* **Node Version:**
6.10.0

* **npm Version:**
3.10.10

**What parser (default, Babel-ESLint, etc.) are you using?**
babel-eslint (v7.1.1)

**Please show your full configuration:**

```
» cat .eslintrc
{
  "parser": "babel-eslint"
}
```

**What did you do? Please include the actual source code causing the issue.**

```js
import React from 'react';
import classNames from 'classnames';

import '../styles/MyComponent.scss';

class MyComponent extends React.Component {
    constructor(props) {
        super(props);

        this.state = { visible: false };
    }

    toggleUserOptions = () => {
        this.setState({ visible: !this.state.visible });
    };

    render() {
        return (<div>/*implementation*/</div>);
    }
}
```

**What did you expect to happen?**

I expected it to pass.

**What actually happened? Please include the actual, raw output from ESLint.**

```
» $(npm bin)/eslint src/resources/webapp/static --cache --quiet --parser=babel-eslint

/Users/dwade/indeed/business-automation/bill-webapp/src/resources/webapp/static/scripts/components/common/UserOptions.js
  14:9   error  Unexpected 'this'  no-invalid-this
  14:46  error  Unexpected 'this'  no-invalid-this

✖ 2 problems (2 errors, 0 warnings)
```

**Comment by not-an-aardvark** [comment]
Thanks for the issue and the pull request. However, we generally don't support experimental syntax in core rules. From our [readme](https://github.com/eslint/eslint/blob/f9ede3fe34105549382b7f94dca1541a6a0eef1c/README.md#what-about-experimental-features):

> ESLint doesn't natively support experimental ECMAScript language features. You can use babel-eslint to use any option available in Babel.

> Once a language feature has been adopted into the ECMAScript standard (stage 4 according to the TC39 process), we will accept issues and pull requests related to the new feature, subject to our contributing guidelines. Until then, please use the appropriate parser and plugin(s) for your experimental feature.

I would recommend using [babel-plugin-eslint](https://github.com/babel/eslint-plugin-babel) to resolve this. It looks like they already have a replacement for the `no-invalid-this` rule to get around this issue.

**Comment by not-an-aardvark** [comment]
Closing this issue -- please use [eslint-plugin-babel](https://github.com/babel/eslint-plugin-babel) for experimental syntax support.
